### PR TITLE
First draft of DI guide

### PR DESCRIPTION
**Pull request by tomdale**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,6 +2,324 @@
...
+start instantiating objects.
+
+As a general rule, this package is designed to be synchronous to achieve maximum
+performance; it is your responsibility to ensure that code is ready before it is
+needed.
+
+One last thing: you may have noticed that the container in this example has both
+a registry and a resolver. The container will look for classes in both, but the
+registry always takes precedence. If the registry is empty, the container will
+fall back to asking the resolver for its help.
```
Comment (locks) [inline README.md]: for help

?

**Comment by dgeb** [comment]
Thank you @tomdale. An excellent start here!

File: README.md [hunk 2]
```diff
@@ -2,6 +2,324 @@
...
 Dependency injection for Glimmer applications.
 
+## What is Dependency Injection?
+
+Dependency injection is a pattern that increases the flexibility, testability
+and consistency of your code.
+
+The three key ideas are:
+
+1. An object's dependencies should be provided to the object when it is created,
```
Comment (tomdale) [inline README.md]: :+1:

File: README.md [hunk 3]
```diff
@@ -2,6 +2,324 @@
...
+HTTP/2 (or even something like a WebSocket), this code is not reusable.
+
+We would have to either duplicate this code, or add some configuration options
+to let the user tell us which protocol they want to use. Of course, that would
+work today, but if we wanted to support HTTP/3 in the future, we'd have to come
+back and add a new configuration option for every new protocol.
+
+What if, instead of telling the server what protocol to use, we could instead
+provide it with an object that encapsulated all of those concerns? Instead of
+having our `HelloWorldServer` import an `HTTPServer` directly, we can provide it
```
Comment (tomdale) [inline README.md]: This reads correctly to my ear (@dgeb?) but I'm open to a clearer way to express this.

File: README.md [hunk 4]
```diff
@@ -2,6 +2,324 @@
...
+with a pre-configured object that implements the same interface. (In this case,
```
Comment (tomdale) [inline README.md]: They are not.

File: README.md [hunk 5]
```diff
@@ -2,6 +2,324 @@
...
+
+```js
+import { Registry } from '@glimmer/di';
+import ProfileComponent from './components/profile';
+
+let registry = new Registry();
+registry.register('component:profile', ProfileComponent);
+```
+
+You probably noticed the string that we're passing to the `register` method:
```
Comment (tomdale) [inline README.md]: ??

File: README.md [hunk 6]
```diff
@@ -2,6 +2,324 @@
...
+`lookup` method, you'll get the same instance of the component:
+
+```js
+let component1 = container.lookup('component:profile');
+let component2 = container.lookup('component:profile');
+
+component1 === component2; // => true
+```
+
+You can create many instances of the same component in one app. So in this case,
```
Comment (tomdale) [inline README.md]: 👍

File: README.md (hunk 1 above)
Comment (tomdale) [inline README.md]: This is grammatically correct. It specifies that the help is coming from the resolver. In "asking the resolver for help" the possession is implicitly the resolver's, but here it's made explicit: "asking the resolver for (its) help."

**Comment by tomdale** [comment]
@locks Updated

File: README.md (hunk 5 above)
Comment (locks) [inline README.md]: I don't remember either :X

**Pull request merged**